### PR TITLE
[hail] teach begin to return its last value

### PR DESCRIPTION
**Pull request by danking**

### CODE REVIEW

**Comment by danking** [comment]
Begin may also be empty, apparently.

File: hail/src/main/scala/is/hail/expr/ir/Emit.scala [hunk 1]
```diff
@@ -766,6 +766,12 @@ class Emit[C](
           }
         }
 
+      case Begin(xs) =>
+        assert(xs.length != 0)
+        xs.init.foreach(emitVoid(_))
+        assert(xs.last.pType != TVoid)
```
Comment (johnc1231) [inline hail/src/main/scala/is/hail/expr/ir/Emit.scala:772]: A `pType` would never equal `TVoid`. Should be `PVoid`? Though I also wonder why it's not ok for `Begin` to end with a `Void` type. Do we mandate that all IR's return something in the end?

File: hail/src/main/scala/is/hail/expr/ir/TypeCheck.scala [hunk 2]
```diff
@@ -346,9 +346,11 @@ object TypeCheck {
       case CombOpValue(i, value, sig) => assert(value.typ == TBinary)
       case _: SerializeAggs =>
       case _: DeserializeAggs =>
-      case x@Begin(xs) =>
-        xs.foreach { x =>
-          assert(x.typ == TVoid)
+      case Begin(xs) =>
+        if (xs.length != 0) {
+          xs.dropRight(1).foreach { x =>
```
Comment (johnc1231) [inline hail/src/main/scala/is/hail/expr/ir/TypeCheck.scala:351]: Why do `dropRight` here when you do `init` everywhere else? is there a difference?

**Comment by danking** [comment]
addressed

**Comment by danking** [comment]
Closing https://hail.zulipchat.com/#narrow/stream/123011-Hail-Dev/topic/Begin

**Pull request closed without merging**